Generate FSR numerical centroids

samuelshaner opened the pull request:
This PR implements the generation of the numerical centroids of the flat source regions. The numerical centroids will be useful in extending OpenMOC to accommodate linear source regions using the methodology implemented in CASMO 5 and published by Rodolfo Ferrer. The formula for computing the centroids was taken from Ferrer's 2012 PHYSOR paper "Linear Source Approximation in CASMO 5". Additionally, the centroids could be useful in implementing strategies to stabilize or accelerate the CMFD acceleration process.

A plotting routine has been added to plot the centroids over the FSRs using the command:

`plotter.plot_centroids(geometry)`

Below is an example of the centroids plot for `simple-lattice.py`:

![centroids](https://cloud.githubusercontent.com/assets/1244537/7426534/4bd9ec14-ef95-11e4-925b-54fc6fc68696.png)

This PR builds on PR #134 and should not be merged until (at least) PR #134 has been merged.

wbinventor commented on sample-input/pin-cell/pin-cell.py:
```diff
@@ -106,6 +106,11 @@
 solver = CPUSolver(geometry, track_generator)
 solver.setNumThreads(num_threads)
 solver.setSourceConvergenceThreshold(tolerance)
+
```
Any reason to use this polar quadrature here?

samuelshaner replied:
Oops, I had that in to test your polar quadrature refactorization. I'll take it out.

wbinventor commented on src/CPUSolver.cpp:
```diff
@@ -185,6 +185,28 @@ void CPUSolver::setNumThreads(int num_threads) {
+ *          ID, sets the Material for each FSR, and assigns a volume based on
+ *          the cumulative length of all of the segments inside the FSR.
+ */
+void CPUSolver::initializeFSRs() {
```
It looks like you already synchronized with my PR #134?

samuelshaner replied:
Correct, I merged with your `fsr-volume-getter` branch so PR #134 should be merged first. If more of your PRs are merged into `develop`, I'll continue to update this PR.

wbinventor commented on src/Geometry.cpp:
```diff
@@ -643,6 +647,27 @@ Point* Geometry::getFSRPoint(int fsr_id) {
+  }
+  catch(std::exception &e) {
+    log_printf(ERROR, "Could not find centroid in FSR: %i. "
+               "Backtrace:%s", fsr_id, e.what());
```
I've tried to remove the backtraces from the error messages, since the logger module throws an exception anyway. This would look like:

```C++
catch(std::exception &e) {
  log_printf(ERROR, "Could not find centroid in FSR: %i.", fsr_id);
```

samuelshaner replied:
Got it. I'll make that fix.

wbinventor commented:
This is a pretty cool PR! I'm going to wait to review the rest of it until after PR #134 has been finalized, since it appears that most of the diffs are from that PR. In reality, I think it might be best to put this in the "back of the line" since some of my other PRs may affect it, in particular #141 and #137. Also, I would recommend merging with (or finalizing) PR #138 and #142 before using the centroids in CMFD or the MOC solve.

samuelshaner commented:
Totally agree that this can wait until PR #137, #138, #141, and #142 are merged. I just haven't had the time to look over those yet and didn't want to go through the process of merging with all the changes of those pending PRs at this time. I'll continue to merge with develop as those PRs are merged in.

wbinventor commented:
I'll review this PR once it has been updated with respect to the current "develop" branch.

samuelshaner commented:
@wbinventor I updated this PR to reflect the changes in the current "develop" branch so it's ready for your review. No big rush if you are busy with other stuff.

wbinventor commented on openmoc/plotter.py:
```diff
@@ -515,6 +515,139 @@ def plot_flat_source_regions(geometry, gridsize=250, xlim=None, ylim=None):
+    x = point.getX()
+    y = point.getY()
+    
+    if xlim is None and ylim is None:
```
Is this conditional actually needed? Will matplotlib actually try to plot a centroid if it is outside the bounds of the original image?

samuelshaner replied:
This conditional is needed in case someone wants to plot only the centroids for a certain region of the geometry. Say someone wants to plot the centroids for one pin cell in an assembly and inputs `xlim` and `ylim` accordingly. If this conditional was not present, the flat source regions would only be plotted for a few the pin cell, but the centroids would be plotted for the entire geometry.

Another option could be to remove the conditional entirely and have the following:

`if xlim is not None:`
`plt.xlim(xlim)`
`if ylim is not None:`
`plt.ylim(ylim)`

What do you think about changing the long condition to just setting the plot limits? The results are the same.

wbinventor replied:
I think whichever option results in the least amount of code is likely the best. But I see your point and don't have an issue with keeping it as is.

wbinventor commented on openmoc/plotter.py:
```diff
@@ -515,6 +515,139 @@ def plot_flat_source_regions(geometry, gridsize=250, xlim=None, ylim=None):
+        centroids_x.append(x)
+        centroids_y.append(y)
+        
+  plt.plot(centroids_x, centroids_y, 'ko', markersize=2)
```
Can we make the centroid marker type and size be keyword arguments to this routine?

samuelshaner replied:
sure

wbinventor commented on openmoc/plotter.py:
```diff
@@ -515,6 +515,139 @@ def plot_flat_source_regions(geometry, gridsize=250, xlim=None, ylim=None):
+        
+  plt.plot(centroids_x, centroids_y, 'ko', markersize=2)
+  
+  plt.title('Centroids')
```
I think the title should be "FSR Centroids" to be more exact perhaps?

samuelshaner replied:
Sure, since we could technically make centroids for cells and lattices.

wbinventor commented on openmoc/plotter.py:
```diff
@@ -515,6 +515,139 @@ def plot_flat_source_regions(geometry, gridsize=250, xlim=None, ylim=None):
+# @param ylim optional list/tuple of the minimim/maximum y-coordinates
+def plot_centroids(geometry, gridsize=250, xlim=None, ylim=None):
+
+  global subdirectory
```
After looking at this routine, is it really necessary? Perahsp an optional ``centroids=False`` keyword param for the ``plot_flat_source_regions(...)`` routine would better for code reuse?

samuelshaner replied:
That's a good idea - I'll update with that.

wbinventor commented on src/Geometry.cpp:
```diff
@@ -613,7 +617,28 @@ Point* Geometry::getFSRPoint(int fsr_id) {
   }
   catch(std::exception &e) {
-    log_printf(ERROR, "Could not find characteristic point in FSR %d", fsr_id);
+    log_printf(ERROR, "Could not find characteristic point in FSR: %i.",
```
Why change from ``%d`` to ``%i`` here?

samuelshaner replied:
Not sure, I'll change it back.

wbinventor commented on src/Geometry.h:
```diff
@@ -143,6 +146,7 @@ class Geometry {
   void setFSRsToKeys(std::vector<std::size_t> FSRs_to_keys);
   void setNumFSRs(int num_fsrs);
   void setCmfd(Cmfd* cmfd);
+  void setFSRCentroid(int fsr, Point* centroid);
```
Should this be public?

samuelshaner replied:
Probably not; I'll make it private.

wbinventor commented on src/Geometry.h:
```diff
@@ -34,9 +34,11 @@ struct fsr_data {
 
-  /** Characteristic point in Universe 0 that lies in FSR */
+  /** Characteristic point in Root Universe that lies in FSR */
   Point* _point;
```
Is it necessary to have both a "characteristic point" and a centroid? It seems now that we have the centroid calculator that we can simply use it as the characteristic point (and just use the ``_centroid`` attribute since it is a more intuitive naming convention).

wbinventor replied:
I just realized that the ``fsr_data`` struct does not have a destructor. This means that the ``_point`` and ``_centroid`` attributes will just be left dangling after an ``fsr_data`` is deleted from the ``Geometry`` (*e.g.*, when running many simulations during a parametric study).

samuelshaner replied:
I thought about that, but there are cases where the centroid can be outside of the FSR (e.g. a ring with no sectors) so it is necessary to have a point that lies within the FSR to construct the linked list of coords that uniquely identifies the FSR. 

I'll add a destructor for the point and centroid.

wbinventor replied:
oh yes that make sense now

wbinventor commented on src/TrackGenerator.cpp:
```diff
@@ -1554,6 +1554,67 @@ void TrackGenerator::correctFSRVolume(int fsr_id, FP_PRECISION fsr_volume) {
+
+  /* Set the centroid for the FSR */
+  for (int r=0; r < num_FSRs; r++){
+    _geometry->setFSRCentroid(r, centroids[r]);
```
Maybe it is just me, but when I see a "setter" routine taking in a pointer, I assume that there are some OO trust boundaries are setup such that the underying class/struct will retain that pointer. Instead, it looks like the ``Geometry::setFSRCentroid(...)`` routine initializes a new ``Point``, copies the data over from the point passed in, and then ``TrackGenerator::generateFSRCentroids()`` deletes the ``Points`` here. 

What about passing in pointers to the ``Points`` as you do here, and simply making each ``fsr_data`` struct point to the very same point in ``Geometry::setFSRCentroid(...)``? And then the ``fsr_data`` destructor will be in charge of deleting each ``Point``.

samuelshaner replied:
That works - I change it to that. I was just thinking it would be better for the Geometry to instantiate its own centroid in case there are changes to OpenMOC in the future where the `setFSRCentroid` method is called with a point allocated on the heap and thus destroyed with the calling method is finished.

wbinventor commented:
@samuelshaner I tried this code out on my machine and it works well. I made a number of comments - once we hash through them I think this is ready to go.

samuelshaner commented:
@wbinventor I updated the PR with based on your comments and made a few other small changes. Here are the notable changes:
- Got rid of `plot_centroids` in favor of boolean in `plot_flat_source_regions` to add centroids on the FSR plot.
- Added destructor for the `fsr_data` struct.
- Modified the handling of `std::vector` and `std::map` objects from the `Geometry` so now they are set/returned by vector instead of by deep copy.
- Added destructor for the surface `_neighbors`.

wbinventor commented on openmoc/plotter.py:
```diff
@@ -442,12 +448,32 @@ def plot_flat_source_regions(geometry, gridsize=250, xlim=None, ylim=None):
+              'marker_size is not an int or float', marker_size)
+
+  if marker_size <= 0:
+    py_printf('ERROR', 'Unable to plot the flat source regions since ' + \
```
remove "since" in this message string (copy and paste error)

wbinventor commented on openmoc/plotter.py:
```diff
@@ -442,12 +448,32 @@ def plot_flat_source_regions(geometry, gridsize=250, xlim=None, ylim=None):
+              'marker_type is not a valid marker (%d)', marker_type)
+
+  if not is_float(marker_size) and not is_integer(marker_size):
+    py_printf('ERROR', 'Unable to plot the flat source regions since ' + \
```
Can the markersize be a float?

samuelshaner replied:
The markersize in Matplotlib is a float, but I think the user should have the option to input it as an int since this is how I have most commonly seen it.

wbinventor replied:
I agree, just wondered since I have always used ints and didn't realize it would accept a float

wbinventor commented on src/Geometry.cpp:
```diff
@@ -613,7 +617,27 @@ Point* Geometry::getFSRPoint(int fsr_id) {
+    point = _FSR_keys_map.at(_FSRs_to_keys.at(fsr_id))._centroid;
+  }
+  catch(std::exception &e) {
+    log_printf(ERROR, "Could not find centroid in FSR: %i.", fsr_id);
```
This is still using "%i" instead of "%d" - I *don't* think this matters, but for consistency throughout the code I think "%d" is better

wbinventor commented on src/Geometry.cpp:
```diff
@@ -1081,6 +1105,10 @@ bool Geometry::withinBounds(LocalCoords* coords){
 }
 
 
+void Geometry::setFSRCentroid(int fsr, Point* centroid){
```
This method needs as Doxygen comment block. And since this is a public method (which I think is dangerous, but I understand why), it should have a ``@details`` block which describes it specifically as a helper routine for the ``TrackGenerator`` and not as something for the user to call from Python.

Speaking of public vs. private methods, we could use the SWIG ``%ignore`` directive to enforce that routines like this one are not exposed within Python. What do you think about this @samuelshaner? This might also apply to the similar discussion we've had with @geogunow about a few questionably public routines needed for the ``TrackGenerator`` and ``Geometry`` to communicate with one another during ray tracing.

samuelshaner replied:
I like that compromise, but it might be wise to consider a cleaner solution in the future.

wbinventor commented on src/Surface.cpp:
```diff
@@ -65,6 +65,14 @@ Surface::Surface(const int id, const char* name){
     delete [] _name;
+
+  if (!_neighbors.empty()){
+    std::map<int, std::vector<Cell*>* >::iterator iter;
```
This would be simpler without the ``for`` loop if we simply use the fact that there are only two integer keys for each halfspace:

```C++
if (!_neighbors.empty()) {
  _neighbors[-1].clear();
  _neighbors[+1].clear();
  _neighbors.clear();
}
```

wbinventor commented on src/TrackGenerator.cpp:
```diff
@@ -1554,6 +1557,64 @@ void TrackGenerator::correctFSRVolume(int fsr_id, FP_PRECISION fsr_volume) {
+
+        x += cos(phi) * curr_segment->_length;
+        y += sin(phi) * curr_segment->_length;
+      }
```
Hmm, I think the way the operations are ordered within this second loop, the final ``segment`` in each ``Track`` will have no contribution to a centroid. Shouldn't the final two lines to increment ``x`` and ``y`` in come before ``centroids[fsr]->setX(...)`` and ``centroids[fsr]->setY(...)```?

samuelshaner replied:
The `x` and `y` variables are initialized to the beginning of the track or equivalently the beginning of the first segment. For each segment, we need to know the segment midpoint, so when looping over the segments the midpoint is found by taking the segment starting point (`x` and `y`) and adding half the segment length times the appropriate sin/cos of the azimuthal angle. `x` and `y` are then incremented at the end of the loop so they become the starting point of the next segment. After the centroid is incremented for the last segment, `x` and `y` are set to the endpoint of the last segment. 

In short, I believe it is correct as is.

wbinventor replied:
Ok, this makes sense now

wbinventor commented:
@samuelshaner - just looked over this a second time - everything looks good save for a few minor comments. I'd like to get your thoughts about using SWIG ``%ignore`` for the public (but not so public) routines created for this PR (and elsewhere in the code), but otherwise am ready to merge this PR.

samuelshaner commented:
Updated the PR with the requested edits.

wbinventor commented on src/Geometry.cpp:
```diff
@@ -1040,9 +1064,9 @@ std::vector<int> Geometry::getFSRsToMaterialIDs() {
  * @param FSR_keys_map map of FSR keys to FSR data
  */
-void Geometry::setFSRKeysMap(std::unordered_map<std::size_t, fsr_data> 
+void Geometry::setFSRKeysMap(std::unordered_map<std::size_t, fsr_data>*
```
Can we add this to the list of SWIG ``%ignore`` routines? I know this isn't a new method, but since it came up in our review here I think it would be good to add.

wbinventor commented on src/Geometry.cpp:
```diff
@@ -1050,8 +1074,8 @@ void Geometry::setFSRKeysMap(std::unordered_map<std::size_t, fsr_data>
  */
-void Geometry::setFSRsToKeys(std::vector<std::size_t> FSRs_to_keys){
-  _FSRs_to_keys = FSRs_to_keys;
+void Geometry::setFSRsToKeys(std::vector<std::size_t>* FSRs_to_keys){
```
ditto SWIG ``%ignore``

wbinventor commented on src/Geometry.cpp:
```diff
@@ -1059,8 +1083,8 @@ void Geometry::setFSRsToKeys(std::vector<std::size_t> FSRs_to_keys){
  */
-void Geometry::setFSRsToMaterialIDs(std::vector<int> FSRs_to_material_IDs){
-  _FSRs_to_material_IDs = FSRs_to_material_IDs;
+void Geometry::setFSRsToMaterialIDs(std::vector<int>* FSRs_to_material_IDs){
```
ditto SWIG ``%ignore``

wbinventor commented on src/Geometry.cpp:
```diff
@@ -1080,6 +1104,22 @@ bool Geometry::withinBounds(LocalCoords* coords){
+ * @param fsr a FSR id
+ * @param centroid a Point representing the FSR centroid
+ */
+void Geometry::setFSRCentroid(int fsr, Point* centroid){
```
ditto SWIG ``%ignore`` unless you think this one is worthwhile to expose in Python

wbinventor commented:
@samuelshaner - I'm ready to approve this once those last three routines are added to the list of SWIG ``%ignore`` routines.

wbinventor commented:
My tests for this PR all pass, but it needs to be synchronized with latest "develop" branch, specifically PR #151, so that I can build and test with Intel.

samuelshaner commented:
Thanks for the comments will. I'm in the Tetons this week with my friend
Andrew, but I'll make the edits as soon as I get a chance.

On Sunday, June 21, 2015, Will Boyd <notifications@github.com> wrote:

> My tests for this PR all pass, but it needs to be synchronized with latest
> "develop" branch, specifically PR #151
> <https://github.com/mit-crpg/OpenMOC/pull/151>, so that I can build and
> test with Intel.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/OpenMOC/pull/144#issuecomment-113941460>.
>


-- 
Regards,
Sam Shaner
Graduate Student
MIT Nuclear Science and Engineering
samuelshaner@gmail.com | (805)748-7684 | 24-607

samuelshaner commented:
@wbinventor I synchronized with the latest "develop" and added the three requested routines to the list of SWIG `%ignore` routines so this PR should be good to merge.
